### PR TITLE
Retrieve issues from both services

### PR DESCRIPTION
**Pull request by LyonesGamer**
Using provided credentials, request a list of GitHub issues updated after a certain time, then the matching list of JIRA issues where they exist. Pair the two lists: for each GitHub issue, if no matching JIRA issue exists, prepare to create it; if an issue exists, prepare to compare the two and update their differences.

Design docs are located at https://docs.google.com/a/coreos.com/document/d/1YjAIqAa-RGC2AZVvlRk1QWbLsJUyt67nIh8lWqKFcK8/edit?usp=sharing.

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -1,61 +1,353 @@
...
+	since, err = time.Parse("2006-01-02", sinceStr)
+	if err != nil {
+		return errors.New("Since date must be in YYYY-MM-DD format")
+	}
+	log.Debug("All config variables are valid!")
+
+	return nil
+}
+
+type Field struct {
```
Comment (kans) [inline cmd/root.go]: Would you rename this to something more descriptive?

File: cmd/root.go [hunk 2]
```diff
@@ -1,61 +1,353 @@
...
+		return err
+	}
+	fields := new([]Field)
+
+	_, err = client.Do(req, fields)
+	if err != nil {
+		return err
+	}
+
+	for _, field := range *fields {
```
Comment (kans) [inline cmd/root.go]: What will happen if we don't set `ghIDFieldID`?

File: cmd/root.go [hunk 3]
```diff
@@ -1,61 +1,353 @@
...
+func compareIssues(ghClient github.Client, jiraClient jira.Client) error {
+	log.Debug("Collecting issues")
+	ctx := context.Background()
+
+	repo := strings.Split(rootCmdCfg.GetString("repo-name"), "/")
+
+	ghIssues, _, err := ghClient.Issues.ListByRepo(ctx, repo[0], repo[1], &github.IssueListByRepoOptions{
+		Since: since,
+		ListOptions: github.ListOptions{
+			PerPage: 100,
```
Comment (kans) [inline cmd/root.go]: Will have to add support for pagination later?

File: cmd/root.go [hunk 4]
```diff
@@ -1,61 +1,353 @@
...
+	log.Debug("Collected all JIRA issues")
+
+	for _, ghIssue := range ghIssues {
+		found := false
+		for _, jIssue := range jiraIssues {
+			id, _ := jIssue.Fields.Unknowns.Int(fmt.Sprintf("customfield_%s", ghIDFieldID))
+			if int64(*ghIssue.ID) == id {
+				found = true
+				if err := updateIssue(*ghIssue, jIssue); err != nil {
+					return err
```
Comment (kans) [inline cmd/root.go]: It would probably be better to log this err and keep trucking - (and the same below).

File: cmd/root.go (hunk 1 above)
Comment (LyonesGamer) [inline cmd/root.go]: Agreed. Maybe `JIRAField`?

File: cmd/root.go (hunk 3 above)
Comment (LyonesGamer) [inline cmd/root.go]: Yes, for now it just operates on the most recent page of results.

**Comment by kans** [comment]
@LyonesGamer 

If you have several more PRs chained after this and don't want to go through rebasing them all multiple times, you could just address the comments in a future PR.

**Comment by LyonesGamer** [comment]
I think it'll be fine; they're all linear, and the comments are mostly small changes so far anyway. They should rebase pretty easily. If there's a comment asking for a big change, I'll probably make it a new PR.

File: cmd/root.go (hunk 2 above)
Comment (LyonesGamer) [inline cmd/root.go]: I added checks that these fields exist on the project. Later I can make some of them optional and have the tool recognize that, but for now they're all required. I'll document it in the README when I write that up.

File: cmd/root.go [hunk 5]
```diff
@@ -1,61 +1,363 @@
...
+
+	var c Config
+	rootCmdCfg.Unmarshal(&c)
+
+	b, err := json.MarshalIndent(c, "", "  ")
+	if err != nil {
+		return err
+	}
+
+	f, err := os.OpenFile(rootCmdCfg.ConfigFileUsed(), os.O_RDWR|os.O_TRUNC|os.O_CREATE, 0751)
```
Comment (squat) [inline cmd/root.go]: why 751? the config is not executable. 644 sounds more appropriate.

File: cmd/root.go [hunk 6]
```diff
@@ -1,61 +1,363 @@
...
 )
 
 var (
-	log             *logrus.Logger
+	log             *logrus.Entry
 	defaultLogLevel = logrus.InfoLevel
 	rootCmdFile     string
 	rootCmdCfg      *viper.Viper
+
+	since               time.Time // The earliest GitHub issue updates we want to retrieve
```
Comment (squat) [inline cmd/root.go]: In general we try to have as few global variables as possible. Why are all of these floating around vs being collected in a config?

Comment (squat) [inline cmd/root.go]: Also, by convention we document types and variables by commenting above the variable: https://blog.golang.org/godoc-documenting-go-code

File: cmd/root.go [hunk 7]
```diff
@@ -1,61 +1,363 @@
...
 	"github.com/fsnotify/fsnotify"
+	"github.com/google/go-github/github"
 	"github.com/spf13/cobra"
 	"github.com/spf13/viper"
+	"golang.org/x/oauth2"
 )
 
 var (
-	log             *logrus.Logger
+	log             *logrus.Entry
```
Comment (squat) [inline cmd/root.go]: Why not logrus.Logger?

**Comment by squat** [comment]
Overall looks ok, Most troubling is the use of lots of global variables rather than collecting all the configuration.

File: cmd/root.go (hunk 7 above)
Comment (LyonesGamer) [inline cmd/root.go]: Entry is a logger with fields attached, such as the app name that's attached in newLogger().

File: cmd/root.go [hunk 8]
```diff
@@ -1,61 +1,363 @@
...
 
 func Execute() {
 	if err := RootCmd.Execute(); err != nil {
-		fmt.Println(err)
+		log.Debugln(err)
 		os.Exit(1)
 	}
 }
 
+func GetGitHubClient(token string) (*github.Client, error) {
```
Comment (LyonesGamer) [inline cmd/root.go]: I admit that I should have been documenting all of these as I went, but I did document them all in #11, if that helps.

File: cmd/root.go [hunk 9]
```diff
@@ -1,61 +1,363 @@
...
+				log.Errorf("Error creating issue for #%d. Error: %s", *ghIssue.Number, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func updateIssue(ghIssue github.Issue, jIssue jira.Issue) error {
+	log.Debugf("Updating JIRA issue %s with GitHub issue %d", jIssue.ID, *ghIssue.ID)
```
Comment (LyonesGamer) [inline cmd/root.go]: Yeah, they're just to help show that compareIssues is working until the function bodies are added in future PRs.

File: cmd/root.go (hunk 5 above)
Comment (LyonesGamer) [inline cmd/root.go]: Oops. You're right. I got my permission values mixed up.

File: cmd/root.go (hunk 8 above)
Comment (squat) [inline cmd/root.go]: no, it means that all the PRs until #11 lands are difficult to reason about. Please break up #11 and spread the documentation into each of the previous PRs

File: cmd/root.go (hunk 7 above)
Comment (squat) [inline cmd/root.go]: Yes I understand that theyre different, what I am asking is why did you change? not a big deal. we typically do not use Entry in other projects but it doesn't at all that it is wrong

Comment (LyonesGamer) [inline cmd/root.go]: It's how it was done in https://github.com/sym3tri/hookz, which is what I was told to use as a template for this project.

Comment (squat) [inline cmd/root.go]: :+1:
